### ISSUE TITLE
Performance extracting values from pointer using chol

### ISSUE THREAD

**Issue by VictorPPB** [cdeterman/gpuR#107]
Hi Charles,

I noticed that extracting values after `chol` reduces time performance when passing a `vclMatrix` or a `gpuMatrix` object. Extracting values makes CPU decomposition faster than gpuR. ¿Is there a way to improve this?

My example is:

```
library(gpuR)
set.seed(1)
N <- 1000000
mm <- matrix(rnorm(N),ncol=sqrt(N),nrow=sqrt(N))
mm <- t(mm)%*%mm
gm <- gpuR::gpuMatrix(mm, type = "float")
gm1 <- gpuR::vclMatrix(mm, type = "float")
system.time(chol(gm))
system.time(chol(gm1))
system.time(chol(gm1)[])
system.time(chol(mm))
```

And I got:

```
> system.time(chol(gm))
   user  system elapsed 
  7.117   0.147   1.048 
> system.time(chol(gm1))
   user  system elapsed 
  0.024   0.024   0.009 
> system.time(chol(gm1)[])
   user  system elapsed 
 14.013   0.295   2.224 
> system.time(chol(mm))
   user  system elapsed 
  0.131   0.001   0.131
```

**Comment by cdeterman** [comment]
Unfortunately there isn't a lot at the moment.  I am aware that the `chol` implementation is rather slow.  This is because such functionality is 'experimental' in the underlying C++ code.  It is implemented simply to have it available and if additional GPU operations need to be applied.  I am intending to create a companion package that links to the `clMagma` library [here](https://github.com/gpuRcore/gpuRclmagma) which should provide a better speedup.  Regardless, device->host and host->device will always be a potential bottleneck.  Overall performance will ultimately come down to how efficiently you process you data on the device (GPU) before moving it back to the host.

Another thought to keep in mind is the device you are using as well.  The better the GPU, the better the performance will be.

**Comment by VictorPPB** [comment]
My cpu opencl device outperforms my gpu. However it keeps slower than simple cpu. 

Anyway, thanks a lot for your time. I hope you can reach a better speedup.

**Comment by cdeterman** [comment]
Out of curiosity, what GPU do you have?

**Comment by VictorPPB** [comment]
There are mi available devices:

```
                                     device device_index device_type
1 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz            0         cpu
2                                  Iris Pro            0         gpu
3        AMD Radeon R9 M370X Compute Engine            0         gpu
```

**Comment by cdeterman** [comment]
So to confirm, when you are doing your benchmarks you are making sure to assign the appropriate context id to make sure the relevant device is actually being used?

For example:

```r
# Use AMD card
gpuMatrix(rnorm(16), nrow = 4, ncol = 4, ctx_id = 3L, type = "float")
```

**Comment by VictorPPB** [comment]
In my example, I used `setContext(i)` for i=1,2,3 to compare times before creating `gm`

````
...
setContext(i)
gm <- gpuR::gpuMatrix(mm, type = "float")
...
````


Which I think is the same thing, isn't it?

**Comment by cdeterman** [comment]
Yes, that is the same for a global assignment instead of individually.  Just wanted to make sure as other users had overlooked that part.
